File: leetcode/top-interview-150/14/Solution.java

/**
 * LONGEST COMMON PREFIX
 * Write a function to find the longest common prefix string amongst an array of
 * strings
 * If there is no common prefix, return an empty string ""
 */

import java.math.*;

class Solution {
    public static String longestCommonPrefix(String[] strs) {
        int i = 0;

        String prfx = strs[i];

        while (i < strs.length - 1) {
            String s1 = prfx;
            String s2 = strs[i + 1];

            int limit = Math.min(s1.length(), s2.length());

            if (s1.length() == 0 || s2.length() == 0)
                return "";

            int k = 0;
            while (k < limit) {
                if (s1.charAt(k) == s2.charAt(k)) {
                    k++;
                } else {
                    break;
                }
            }

            prfx = s1.substring(0, k);

            i++;

        }
        return prfx;
    }

    public static void main(String[] args) {
        String[] strs = { "flower" };
        String rs = longestCommonPrefix(strs);
        System.out.println(rs);
    }

}

File: leetcode/top-interview-150/14/Solution.go
package main

import (
	"fmt"
	// "strings"
)

func main() {
	strs := []string{"flower", "flow", "flight"}
	// strs := []string{"dog", "racecar", "car"}
	// fmt.Print(longestCommonPrefix(strs))
	fmt.Print(longestCommonPrefix_reducePrefix(strs))
}

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	for i := 0; i < len(strs[0]); i++ {
		char := strs[0][i] // The character of index i of the first string element

		// Compare current character with the same index in every string of strs
		for _, s := range strs {
			// if i exceeds the length of current string -> the current string is shorter than current string -> then return
			// if character is not match then return
			if i >= len(s) || s[i] != char {
				return strs[0][:i]
			}
		}
	}
	// If no match was found, the entire first element string is the common prefix
	return strs[0]
}

func longestCommonPrefix_reducePrefix(strs []string) string {
	prefix := strs[0]

	for i := 1; i < len(strs); i++ {
		s := strs[i]
		// if s not start with prefix then prefix - 1 last character
		// for !strings.HasPrefix(s, prefix) {
		// 	prefix = prefix[:len(prefix)-1]
		// }
		for len(s) < len(prefix) || s[:len(prefix)] != prefix {
			prefix = prefix[:len(prefix)-1]
		}
	}
	return prefix
}
